### PR TITLE
Sh/pkg install

### PR DESCRIPTION
**Pull request by shetzel**
Adds the ability to install a 2GP package (04t).

@W-10832794@

I manually tested:
1. installing an unlocked package; no polling
2. installing an unlocked package; no polling, --json
3. installing a managed package; no polling
4. installing a managed package; no polling, --json
5. installing an unlocked package; with polling
6. installing an unlocked package; with polling, --json
7. installing a managed package; with polling
8. installing a managed package; with polling, --json
9. installing an unlocked package that errors
10. installing an unlocked package that errors, --json
11. installing a package with `upgradeType = Delete` prompts for confirmation
12. installing a package and waiting for publish
13. installing a managed package with non-default `upgradeType` displays warning message
14. installing a managed package with non-default `apexCompileType` displays warning message
15. installing a package with invalid package ID length throws error
16. installing a package with invalid package ID prefix throws error
17. installing a package with a package alias outside of a project throws error
18. installing a package with a package alias not in sfdx-project.json throws error

I did not test:
1. installing a package with external sites (should prompt the user for confirmation)
2. installing a package that was not yet replicated to the install org (waitForPublish error condition)
3. installing a package that needs an installation key

### CODE REVIEW

File: src/package/packageInstall.ts [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    if (request.ApexCompileType !== defaults.ApexCompileType) {
+      const msg = installMsgs.getMessage('apexCompileOnlyForUnlockedWarning');
+      await Lifecycle.getInstance().emit('PackageInstallRequest:warning', msg);
+      delete request.ApexCompileType;
+    }
+  }
+
+  await Lifecycle.getInstance().emit('PackageInstallRequest:presend', request);
+
+  const result = await connection.tooling.create('PackageInstallRequest', request);
```
Comment (peternhale) [inline src/package/packageInstall.ts:73]: Do we want to catch unhandled exceptions here to provide better messaging or emit a event?

File: src/package/packageInstall.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+  await Lifecycle.getInstance().emit('PackageInstallRequest:postsend', result);
+
+  const packageInstallRequestId = result.id;
+  if (!packageInstallRequestId) {
```
Comment (peternhale) [inline src/package/packageInstall.ts:78]: Good place for
``` javascript
const packageInstallRequestId = Optional.of(result.id).orElseThrow(() => 
  installMsgs.createError('packageInstallRequestError', [
        request.SubscriberPackageVersionKey,
        result.errors.toString(),
  ])
);

File: src/package/packageInstall.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+    const sites = [...rssUrls, ...cspUrls];
+    if (sites.length) {
+      return sites;
+    }
+  }
+}
+
+async function getStatus(connection: Connection, installRequestId: string): Promise<PackageInstallRequest> {
+  const result = await connection.tooling.retrieve('PackageInstallRequest', installRequestId);
```
Comment (peternhale) [inline src/package/packageInstall.ts:136]: I know retrieve doesn't apply the type correctly. maybe replace with singleRecordQuery?

File: src/package/packageInstall.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    error.setData(packageInstallRequest);
+    if (error.stack && e.stack) {
+      // add the original stack to this new error
+      error.stack += `\nDUE TO:\n${e.stack}`;
+    }
+    throw error;
+  }
+}
+
+export async function waitForPublish(
```
Comment (peternhale) [inline src/package/packageInstall.ts:198]: Why is this name `waitForPublish`?

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/package/packageInstall.ts (hunk 4 above)
Comment (shetzel) [reply src/package/packageInstall.ts:198]: To match the associated command flag and some of the help text around it.  At first I had it as `waitForAvailability` but changed it so it was clearer that it corresponded to the command flag.

File: src/package/packageInstall.ts (hunk 3 above)
Comment (shetzel) [reply src/package/packageInstall.ts:136]: Maybe.  Whatever we decide we should be consistent.  I was hoping we could modify jsforce to return proper typing.

File: src/package/packageInstall.ts (hunk 1 above)
Comment (shetzel) [reply src/package/packageInstall.ts:73]: I don't know.  This matches the original behavior.  I wouldn't know what errors to catch to provide better messaging.

File: src/package/package.ts [hunk 5]
```diff
@@ -24,6 +39,29 @@ export class Package extends AsyncCreatable<PackageOptions> implements IPackage
...
+   *
+   * @param id Salesforce ID for a specific package resource
+   * @param type The type of package ID
+   */
+  public static validateId(id: string, type: PackageIdType): void {
+    const prefix = packagePrefixes[type];
+    if (!id.startsWith(prefix)) {
+      throw messages.createError('invalidPackageId', [type, id, prefix]);
+    }
+    if (![15, 18].includes(id.length)) {
```
Comment (peternhale) [inline src/package/package.ts:60]: Use sfdx-core `validateSalesforceId`
``` suggestion
    if (!validateSalesforceId(id)) {
```

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/package/package.ts (hunk 5 above)
Comment (shetzel) [reply src/package/package.ts:60]: Thanks!  I didn't know about that.

**Review by peternhale — APPROVED** [review]

**Review by WillieRuemmele — APPROVED** [review]

**Pull request merged**